### PR TITLE
feat: add /call command to TUI to execute eth rpc methods

### PR DESCRIPTION
**Pull request by richard-ramos**
Adds a new command `/call` or `/rpc` that receives a json string containing the call being sent to an ethereum provider
image

### CODE REVIEW

File: examples/client/tui/commands.nim
Comment (siphiuel) [inline examples/client/tui/commands.nim:734]: Not sure if this is better, but it's possible to write `let input = if args.len > 0: args[0] else: ""`

**Review by siphiuel — APPROVED** [review]

File: examples/client/client/tasks.nim
Comment (emizzle) [inline examples/client/client/tasks.nim:1051]:
```suggestion
proc callRPC*(input: JsonNode) {.task(kind=no_rts, stoppable=false).} =
```

File: examples/client/tui/common.nim
Comment (emizzle) [inline examples/client/tui/common.nim:128]:
```suggestion
  CallRpc* = ref object of Command
    input*: JsonNode
```

File: status/api/provider.nim
Comment (emizzle) [inline status/api/provider.nim:10]:
```suggestion
proc callRPC*(self: StatusObject, input: JsonNode): CallRPCResult  {.raises:
```

File: examples/client/tui/commands.nim
Comment (emizzle) [inline examples/client/tui/commands.nim:738]: Maybe we could have `input` be `JsonNode`, such that we are forced to validate correct JSON input here. Especially if the user is typing in JSON 😳  If the JSON can't be parsed or is invalid, it would prevent a round-trip through the task runner.

**Review by emizzle — CHANGES_REQUESTED** [review]
Overall, really nice work @richard-ramos 🤩

### General comment
As a general comment, given that this PR implements only part of send transaction, I think that we could treat this PR in one of two ways:
1. as a draft until `sendTransaction` is included
2. as a PR but without `sendTransaction` and follow up with a `sendTransaction` PR

### Using ethcallsigs
I really like the use the use of `nim-web3`, well done.

For calling RPC methods, constructing JSON calls and then pasting them in to the TUI feels like it is a bit on the unfriendly side, but I do see how this functionality could be useful for debugging in the future. I guess in an ideal world, it could potentially be more user-friendly to parse parameters based on the method. As an example, if we wanted to get a block by block number, `eth_getBlockByNumber`, we could do something like:
```shell
# tui input
/call eth_getBlockByNumber 0x123 true
```
These parameters could then be parsed and plugged in to the appropriate [ethcallsig](https://github.com/status-im/nim-web3/blob/master/web3/ethcallsigs.nim), with their usage being like so:
```
# status/api/provider.nim
let block = await self.web3.provider.eth_getBlockByNumber(blockId, fullTransactions)
```
The parsing for each call would have to be different. Compared to `eth_getBlockByNumber`, `eth_getStorageAt` needs three different parameters:
```shell
# tui input
/call eth_getStorageAt 0xBdc1F0460b38876a82Fe8416210DbC49e28c6298 10 0x1234
```
```
# status/api/provider.nim
let block = await self.web3.provider.eth_getStorageAt(address, quantity, blockId)
```
It's worth noting that in other parts of `nim-status` (future additions, out of scope of this PR), if we need to call RPC methods, we should be using the [ethcallsigs](https://github.com/status-im/nim-web3/blob/master/web3/ethcallsigs.nim) instead of unnecessarily constructing JSON objects that skirt around compiler support. Likely, these use cases would not need to be abstracted out in to a generalised call like we are doing for `callRpc`, so hopefully should be much easier to implement ad-hoc.

### Cosmetic preferences
To summarise a general comment about casing, I prefer to treat acronyms like any other word and therefore use camel casing rules, ie I prefer `callRpcEvent` as opposed to `callRPCEvent`. Any usage of `RPC` would be `rpc`, unless it is not at the start, in which case it would be `Rpc`. Of course, this is only my opinion though. IMO, it seems for long acronyms like `secp256k1`, it could feel like someone was swearing at you `SECP256K1` 🤬 😂

**Comment by iurimatias** [comment]
> Overall, really nice work @richard-ramos 🤩
> 
> ### General comment
> As a general comment, given that this PR implements only part of send transaction, I think that we could treat this PR in one of two ways:
> 
> 1. as a draft until `sendTransaction` is included
> 2. as a PR but without `sendTransaction` and follow up with a `sendTransaction` PR
> 
> ### Using ethcallsigs
> I really like the use the use of `nim-web3`, well done.
> 
> For calling RPC methods, constructing JSON calls and then pasting them in to the TUI feels like it is a bit on the unfriendly side, but I do see how this functionality could be useful for debugging in the future. I guess in an ideal world, it could potentially be more user-friendly to parse parameters based on the method. As an example, if we wanted to get a block by block number, `eth_getBlockByNumber`, we could do something like:
> 
> ```shell
> # tui input
> /call eth_getBlockByNumber 0x123 true
> ```
> 
> These parameters could then be parsed, validated, and plugged in to the appropriate [ethcallsig](https://github.com/status-im/nim-web3/blob/master/web3/ethcallsigs.nim), with their usage being like so:
> 
> ```
> # status/api/provider.nim
> let block = await self.web3.provider.eth_getBlockByNumber(blockId, fullTransactions)
> ```
> 
> The parsing and validating for each call would have to be different due to the different parameters. Compared to `eth_getBlockByNumber`, `eth_getStorageAt` needs three different parameters:
> 
> ```shell
> # tui input
> /call eth_getStorageAt 0xBdc1F0460b38876a82Fe8416210DbC49e28c6298 10 0x1234
> ```
> 
> ```
> # status/api/provider.nim
> let block = await self.web3.provider.eth_getStorageAt(address, quantity, blockId)
> ```
> 
> But it feels like when we are allowing free text, we could be validating parameters for better UX and then ultimately that would provide for better compiler support with strongly-typed access.
> 
> It's worth noting that in other parts of `nim-status` (future additions, out of scope of this PR), if we need to call RPC methods, we should be using the [ethcallsigs](https://github.com/status-im/nim-web3/blob/master/web3/ethcallsigs.nim) instead of unnecessarily constructing JSON objects that skirt around compiler support. Likely, these use cases would not need to be abstracted out in to a generalised call like we are doing for `callRpc`, so hopefully should be much easier to implement ad-hoc.
> 
> ### Cosmetic preferences
> To summarise a general comment about casing, I find it preferable to treat acronyms like any other word and therefore use camel casing rules, ie using `callRpcEvent` as opposed to `callRPCEvent`. Any usage of `RPC` would be `rpc`, unless it is not at the start, in which case it would be `Rpc`. Of course, this is only my opinion though.

- we should avoid drafts if we can, specially in a project like this, it risks having a hard rebase needed, better to do in a separate PR
- the unfriendly TUI thing is not something we need to be worried about and it's not worth the work at this stage, we're building a library and the TUI is there just to help but it's not meant to be a user friendly client (not at this stage anyway)

File: status/callrpc.nim
Comment (richard-ramos) [reply status/callrpc.nim:55]: This is done this way so I can pass the result of `settings.getNetwork(networkName)` and `settings.getCurrentNetwork()` directly to the `getWeb3Conn` and have this proc check if the network contains a value,  opossed to having this verification done in both `newWeb3` procs. 
Also `getWeb3Conn` is not a procedure that can be exported. it's a helper for `newWeb3`. Perhaps it should have a different name to avoid this confusion?

File: status/networks.nim
Comment (richard-ramos) [reply status/networks.nim:27]: It should be okay. I copied it from https://github.com/status-im/status-desktop/blob/master/Makefile#L220

File: status/networks.nim
Comment (richard-ramos) [reply status/networks.nim:44]: `https://` was not working. I had to switch to websockets.

File: examples/client/tui/commands.nim
Comment (richard-ramos) [reply examples/client/tui/commands.nim:772]: Yeah. Sounds like a good idea

**Comment by richard-ramos** [comment]
I removed all the code related to sending transactions because it will be part of a separate PR.

I also did most of the required changes but in relation to what was asked in "Using ethcallsigs", I decided to implement it differently: The call command now receives two parameters, with the last one being optional, line in these examples
```
/call eth_gasPrice
/call eth_getBlockByNumber ["0x5BAD55",false]
```

While I agree that using ethcallsigs make sense to use in other parts of the project, I do think that for the TUI, implementing it this way should be good enough for now to achieve a middle point between introducing a JSON which is not great for UX, and having validation per parameters depending on the RPC method being called which would probably require to allocate a lot of time for something that IMO will not be used IRL

File: examples/client/tui/commands.nim
Comment (emizzle) [inline examples/client/tui/commands.nim:756]:
```suggestion
        "params cannot be blank and must be valid JSON, please provide an input.")
```

Comment (emizzle) [reply examples/client/tui/commands.nim:756]: This is because if there was a JSON parse error in `new` (ie user entered bad json), then this will be shown.

File: status/callrpc.nim
Comment (emizzle) [reply status/callrpc.nim:55]: Ah I see, I had forgotten that `getNetwork` and `getCurrentNetwork` returned `Option[Network]` 👍

File: status/networks.nim
Comment (emizzle) [reply status/networks.nim:44]: Oh that's very interesting. Is infura removing support for https? (no need to answer if you don't know, just asking out loud).

**Review by emizzle — APPROVED** [review]
Awesome work, Richard! Thanks for all the updates 💪

**Comment by emizzle** [comment]
@iurimatias 

> we should avoid drafts if we can, specially in a project like this, it risks having a hard rebase needed, better to do in a separate PR

In my opinion, draft PRs are a great way for a developer to welcome feedback on a PR that isn't quite polished and ready for merging. This can help ensure they are going in the right direction before the time is spent cleaning things up and getting it ready to merge. What is a hard rebase?

> the unfriendly TUI thing is not something we need to be worried about and it's not worth the work at this stage, we're building a library and the TUI is there just to help but it's not meant to be a user friendly client (not at this stage anyway)

I understand and definitely agree with not wanting to put too much effort in to the TUI. I wasn't really expecting the work on the TUI to get done; a lot of it was commentary on idealities, but also a note on how we could move forward when calling RPC methods *from within nim-status* in the future.  I think the `JsonNode` input parameters is a good solution for the `callRpc` API proc. And richard's changes to the tui were a perfect middle ground effort.

File: examples/client/tui/actions.nim
Comment (michaelsbradleyjr) [inline examples/client/tui/actions.nim:415]: Not a big deal, but it's nice if the name in the comment matches the type name.

Actually, at this point it probably makes to create an `acitons/` dir and start splitting actions into separate modules. That can be done in a separate PR, of course; I'm just thinking aloud.

File: examples/client/tui/commands.nim
Comment (michaelsbradleyjr) [inline examples/client/tui/commands.nim:739]: The user might have typed in valid json that contained spaces, e.g. between `"key":` and the following value.

So here I think `args[1..^1]` should be joined on single-space (`" "`) and then parsed.

Also, I think we should have an `else` case for `if args.len > 1` where params is set to the empty string.

Then, in `proc command` below for `command: CallRpc`, the situation re: blank params and invalid JSON can be separated.

File: examples/client/tui/commands.nim
Comment (michaelsbradleyjr) [inline examples/client/tui/commands.nim:756]: See my previous comment re: separating `params cannot be blank` and `valid JSON`.

File: examples/client/tui/common.nim
Comment (michaelsbradleyjr) [inline examples/client/tui/common.nim:201]:
```suggestion
    "call": @["rpc"],
```

File: status/api.nim
Comment (michaelsbradleyjr) [inline status/api.nim:2]: Again, it's nice to consistently alphabetize.

File: status/api.nim
Comment (michaelsbradleyjr) [inline status/api.nim:4]: Here also.

File: status/api/common.nim
Comment (michaelsbradleyjr) [inline status/api/common.nim:10]: Here also.

File: status/api/provider.nim
Comment (michaelsbradleyjr) [inline status/api/provider.nim:5]: If there's only one, the `[]` can be left out, but no big deal.

File: status/api/common.nim
Comment (michaelsbradleyjr) [inline status/api/common.nim:63]:
```suggestion
proc web3*(self: StatusObject, network: string): Web3 {.raises: [Defect, Exception]} =
```

We should indicate `Defect` in the `{.raises.}` pragma.

It would better to not declare `Exception` in the `{.raises.}` pragma, as it somewhat defeats the purpose (even if it makes the compiler happy). What we really want is to declare specific kinds of exceptions, e.g. `KeyError` or maybe error/s that getSettings could raise.

However, there's a big refactor @emizzle is working on (almost done) related to this (in a general sense), so it's something that can be sorted out in a future PR if [`KeyError`](https://nim-lang.org/docs/system.html#KeyError) isn't enough.

File: status/private/callrpc.nim
Comment (michaelsbradleyjr) [inline status/private/callrpc.nim:63]:
```suggestion
    raise (ref Web3Error)(msg: "config not found for network " & networkName)
```

File: status/private/networks.nim
Comment (michaelsbradleyjr) [inline status/private/networks.nim:28]: We can eventually supply the default via the Makefile, like we do in status-desktop.

**Review by michaelsbradleyjr — CHANGES_REQUESTED** [review]
Really nice work!

I've offered a few stylistic points below, but they're not blockers for this PR.

The two blockers concern (1) my comments re: parsing of the JSON supplied by the user via the TUI, and (2) the mistake in the `aliased` table.

**Comment by michaelsbradleyjr** [comment]
> the unfriendly TUI thing is not something we need to be worried about and it's not worth the work at this stage, we're building a library and the TUI is there just to help but it's not meant to be a user friendly client (not at this stage anyway)

I agree, the focus should be on dogfooding and *not* great UI/UX.

But, echoing @emizzle's sentiments, I think we can strike a balance between landing too-poor UI/UX and spending unwarranted time on UI/UX. To my mind, @richard-ramos performed an *awesome* balancing act in this PR.

> we should avoid drafts if we can, specially in a project like this, it risks having a hard rebase needed, better to do in a separate PR

A few thoughts, and some *only-my-opinions* are definitely included:

This week it would have been quite difficult for one PR or another to avoid a rebase. We've not been moving stuff around for the sake of it, but trying to make the foundations more solid, and some of those efforts have been happening in parallel, so any one PR would likely have to take the changes into account.

The reorgs/refactors won't go on forever, we're very close to having the foundation we need and want.

Re: this repository and its current GHA workflows, putting a PR in *Draft* is a way for the author to signal:
* I'm want this branch to be built/tested on all three platforms (Linux, macOS, Windows).
* I'm inviting review of the work I've done, which I feel needs some feedback before it's ready to merge.

We quite recently ran into a situation where Eric and I decided to merge one of @siphiuel's PRs that wasn't ready yet, though that wasn't clear to us. While we could use an approach of author/other marking a PR with a red-colored `don't merge yet` label, putting a PR into draft seems like a more elegant solution, since draft PRs simply cannot be merged.

Some of us (sometimes) build on all three platforms locally, in the context of PR authoring, but most of us don't or don't do it all the time. We *could* change nim-status' GHA workflows so they run on any commit to any branch, but we don't have an unlimited number of GHA minutes per month (for the status-im org), so limiting GHA runs to PRs and the `master` branch seems like a good idea.

File: examples/client/client.nim
Comment (emizzle) [reply examples/client/client.nim:122]: There is actually already an enum in `status/callrpc.nim`: `RemoteMethod`. Maybe we could use that?

File: examples/client/tui/commands.nim
Comment (michaelsbradleyjr) [reply examples/client/tui/commands.nim:739]: > `else` case for if `args.len > 1` where params is set to the empty string.

Whoops, looking over it again I see params needs to be of type `JsonNode`. So there's no need for an `else` case, but we could still separate the cases w.r.t. checking if it's JNull else checking the `len` of JArray being `> 0`.

**Review by michaelsbradleyjr — APPROVED** [review]

**Comment by michaelsbradleyjr** [comment]
@richard-ramos I just rebased this PR against current `master`. In the process of doing so, I realized that nim-status' `callRpc` is synchronous but it should be async.

So I introduced and squashed-in changes that make `callRpc` async. I was careful to review the rebased PR + my changes relative to the changes the original PR made on top of 1adb3045d66add8df44dad7a21cd1fb8e0be4dcf. Tests pass, the client builds and `/callRpc` works as expected. Even so, you may want to take a look to check that I didn't mess something up in the process.

In `api/provider` there's a `raises: [Exception]`. Now that #265 has landed, we should probably take a different approach for that. @emizzle, could you take a look?

**Comment by michaelsbradleyjr** [comment]
I've temporarily added a `don't merge yet` label until Richard has had time to check the changes I introduced and we together decided what we want to do re: the `raises: [Exception]`.

**Comment by emizzle** [comment]
@michaelsbradleyjr @richard-ramos, I've updated (and squashed-in) this brach to fix the `{.raises: [Exception].}.` The `api/proivder.nim` module didn't have a `{.push raises: [Defect].}` at the top of the module, so the `{.raises: [Exception].}.` likely was satisfying the compiler's analysis that a `Defect` could have been raised. 

Fixed by adding `{.push raises: [Defect].}` and removing the `{.raises.}` pragma from the proc (effectively equivalent to `{.raises: [Defect].}`).

**Pull request merged**
